File: Data Structures & Algorithms/lru-cache/submission-2.go
type Node struct {
    key, val int
    next, prev *Node
}

func NewNode(key,val int) *Node {
    return &Node{key:key,val:val,next:nil,prev:nil}
}

type LRUCache struct {
    cache map[int]*Node
    head, tail *Node
    capacity int
}

func Constructor(capacity int) LRUCache {
    res := LRUCache {
        cache: map[int]*Node {},
        head: NewNode(0,0),
        tail: NewNode(0,0),
        capacity: capacity,
    }
    res.head.next = res.tail
    res.tail.prev = res.head
    return res
}

func (this *LRUCache) remove(node *Node) {
    next,prev := node.next, node.prev
    next.prev, prev.next = prev, next
}

func (this *LRUCache) insert(node *Node) {
    next, prev := this.tail, this.tail.prev
    node.next, node.prev = next, prev
    next.prev = node
    prev.next = node
}

func (this *LRUCache) Get(key int) int {
    node, ok := this.cache[key]
    if ok {
        this.remove(node)
        this.insert(node)
        return node.val
    }
    return -1
}

func (this *LRUCache) Put(key int, value int) {
    node, ok := this.cache[key]
    if ok {
        this.remove(node)
    }
    this.cache[key] = NewNode(key,value)
    this.insert(this.cache[key])

    if len(this.cache) > this.capacity {
        lru := this.head.next
        this.remove(lru)
        delete( this.cache, lru.key )
    }
}

File: Data Structures & Algorithms/queue/submission-2.py
class Node:
    def __init__(self,val:int,next=None,prev=None):
        self.val = val
        self.next = next
        self.prev = prev

class Deque:
    def __init__(self):
        self.head = Node(0)
        self.tail = Node(0)
        self.head.next,self.tail.prev = self.tail,self.head

    def isEmpty(self) -> bool:
        return self.head.next == self.tail

    def append(self, value: int) -> None:
        prev = self.tail.prev
        self.tail.prev = Node(value,self.tail,self.tail.prev)
        prev.next = self.tail.prev

    def appendleft(self, value: int) -> None:
        next = self.head.next
        self.head.next = Node(value,self.head.next,self.head)
        next.prev = self.head.next

    def pop(self) -> int:
        if not self.tail.prev.prev: return -1
        node = self.tail.prev
        self.tail.prev = self.tail.prev.prev
        self.tail.prev.next = self.tail
        return node.val

    def popleft(self) -> int:
        if not self.head.next.next: return -1
        node = self.head.next
        self.head.next = self.head.next.next
        self.head.next.prev = self.head
        return node.val